src/music/dmusic.cpp

```diff
@@ -19,86 +19,1095 @@
+		bool operator ==(long offset) const {
+			return this->file_offset == offset;
+		}
+	};
+
+	std::vector<DLSInstrument> instruments;
+	std::vector<POOLCUE> pool_cues;
+	std::vector<DLSWave> waves;
+
+	/** Try loading a DLS file into memory. */
+	bool LoadFile(const TCHAR *file);
+
+private:
+	/** Load an articulation structure from a DLS file. */
+	bool ReadDLSArticulation(FILE *f, DWORD list_length, std::vector<CONNECTION> &out);
+	/** Load a list of regions from a DLS file. */
+	bool ReadDLSRegionList(FILE *f, DWORD list_length, DLSInstrument &instrument);
+	/** Load a single region from a DLS file. */
+	bool ReadDLSRegion(FILE *f, DWORD list_length, std::vector<DLSRegion> &out);
+	/** Load a list of instruments from a DLS file. */
+	bool ReadDLSInstrumentList(FILE *f, DWORD list_length);
+	/** Load a single instrument from a DLS file. */
+	bool ReadDLSInstrument(FILE *f, DWORD list_length);
+	/** Load a list of waves from a DLS file. */
+	bool ReadDLSWaveList(FILE *f, DWORD list_length);
+	/** Load a single wave from a DLS file. */
+	bool ReadDLSWave(FILE *f, DWORD list_length, long offset);
 };
 
-static ProcPtrs proc;
+#pragma pack(2)
+/** A RIFF chunk header. */
+struct ChunkHeader {
+	FOURCC type;  ///< Chunk type.
+	DWORD length; ///< Length of the chunk, not including the chunk header itself.
+};
 
+/** Buffer format for a DLS wave download. */
+struct WAVE_DOWNLOAD {
+	DMUS_DOWNLOADINFO   dlInfo;
+	ULONG               ulOffsetTable[2];
+	DMUS_WAVE           dmWave;
+	DMUS_WAVEDATA       dmWaveData;
+};
+#pragma pack()
 
-const char *MusicDriver_DMusic::Start(const char * const *parm)
+struct PlaybackSegment {
+	uint32 start, end;
+	size_t start_block;
+	bool loop;
+};
+
+static struct {
+	bool shutdown;    ///< flag to indicate playback thread shutdown
+	bool playing;     ///< flag indicating that playback is active
+	bool do_start;    ///< flag for starting playback of next_file at next opportunity
+	bool do_stop;     ///< flag for stopping playback at next opportunity
+
+	int preload_time; ///< preload time for music blocks.
+	byte new_volume;  ///< volume setting to change to
+
+	MidiFile next_file;           ///< upcoming file to play
+	PlaybackSegment next_segment; ///< segment info for upcoming file
+} _playback;
+
+/** Handle to our worker thread. */
+static ThreadObject *_dmusic_thread = NULL;
+/** Event to signal the thread that it should look at a state change. */
+static HANDLE _thread_event = NULL;
+/** Lock access to playback data that is not thread-safe. */
+static ThreadMutex *_thread_mutex = NULL;
+
+/** The direct music object manages buffers and ports. */
+static IDirectMusic *_music = NULL;
+/** The port object lets us send MIDI data to the synthesizer. */
+static IDirectMusicPort *_port = NULL;
+/** The buffer object collects the data to sent. */
+static IDirectMusicBuffer *_buffer = NULL;
+/** List of downloaded DLS instruments. */
+static std::vector<IDirectMusicDownload *> _dls_downloads;
```

Comment:
Do these globals need to be separate from the _playback variable?

Reply:
Do they need to? No. The DMusic interfaces are global in the current driver, so I just went with that.